### PR TITLE
allow statics pointing to mutable statics

### LINKED ISSUE #120450: Mutable reference protected by a mutex in static context is considered UB
I tried this code:

```rust
#![feature(const_mut_refs)]

use std::sync::Mutex;

static MUTABLE_REFERENCE_HOLDER: Mutex<&mut [u8]> = Mutex::new(&mut []);

fn main() {}
```

I expected to see this happen: This code should compile because the mutable reference is protected by a Mutex and therefore, it is safe to access.

Instead, this happened: I receive the following compiler error:
```
error[E0080]: it is undefined behavior to use this value
 --> src/main.rs:5:1
  |
5 | static MUTABLE_REFERENCE_HOLDER: Mutex<&mut [u8]> = Mutex::new(&mut []);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ constructing invalid value at .data.value: encountered mutable reference in a `const` or `static`
  |
  = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rustc repository if you believe it should not be considered undefined behavior.
  = note: the raw bytes of the constant (size: 24, align: 8) {
              0x00 │ 00 00 00 00 00 __ __ __ ╾───────alloc3────────╼ │ .....░░░╾──────╼
              0x10 │ 00 00 00 00 00 00 00 00                         │ ........
          }
```

I am not entirely sure if I'm missing something here. Is this UB or not?
Since a mutable reference to a slice of u8 is Send the mutex is Send + Sync. Therefore, I assume this should be allowed? 

If this is in fact a mistake in the compiler, I'd be glad to try to provide a pull request with a fix. In my head the compiler need to look at the Sync trait and if it is implement by a type, this is not UB.

### Meta

`rustc --version --verbose`:
```
rustc 1.77.0-nightly (6b4f1c5e7 2024-01-27)
binary: rustc
commit-hash: 6b4f1c5e782c72a047a23e922decd33e7d462345
commit-date: 2024-01-27
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/120450 for good. We can even simplify our checks: no need to specifically go looking for mutable references in const, we can just reject any reference that points to something mutable.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/validity.rs [hunk 1]
```diff
@@ -511,20 +503,19 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
...
-                        && self.ctfe_mode.is_some_and(|c| !c.may_contain_mutable_ref())
-                    {
-                        throw_validation_failure!(self.path, MutableRefInConstOrStatic);
-                    }
-                    if alloc_actual_mutbl == Mutability::Mut
-                        && matches!(self.ctfe_mode, Some(CtfeValidationMode::Const { .. }))
-                    {
-                        throw_validation_failure!(self.path, ConstRefToMutable);
+                    // In a const, everything must be completely immutable.
+                    if matches!(self.ctfe_mode, Some(CtfeValidationMode::Const { .. })) {
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/interpret/validity.rs:513]: So we don't have any tests testing the promoted code path? Or is that rejected elsewhere even in unleash mode?

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/validity.rs:513]: With promoteds we have to accept this or else we fail [this recently added test](https://github.com/rust-lang/rust/blob/408eeae59d35cbcaab2cfb345d24373954e74fc5/tests/ui/consts/static-promoted-to-mutable-static.rs).

Comment (oli-obk) [reply compiler/rustc_const_eval/src/interpret/validity.rs:513]: 🤣 

> We know from the internet

(sorry, I know what you meant, but I literally lolled)

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/validity.rs:513]: :joy: 

(For anyone else reading along, I meant "interner".)

**Comment by workingjubilee** [comment]
I am thinking about this change from a teaching perspective. I have a small question: If I am understanding this correctly, previously we had "const contexts" which use a set of const evaluation rules to produce a value, and this basically cleaves the world somewhat into "const contexts" and "static contexts", where the rules are nominally shared but some final values become permitted that were not sensical previously. Yes?

**Comment by RalfJung** [comment]
const and static always had a bunch of subtle differences, so this distinction is not entirely new. I would still call them all "const contexts" but there are some particularities that only apply to some const contexts.

For example, `const`/`const fn` are not allowed to mention statics, but `static` is allowed to mention other statics (and even itself).

**Pull request merged**